### DISCUSSION TITLE
Clueless: Administration of a central CUPS server by HTTPS and admin_uri/SERVER_NAME

### DISCUSSION THREAD

**Discussion by noseshimself** [OpenPrinting/cups#678] [Q&A]
I spent a day now trying several approaches but do not see where I'm wrong.

I'm running a central print server (to make it more difficult in a docker container) connected to a (Nebula) mesh overlay network. Printing is not a problem but doing the administration is...

I set ServerName in cupsd.conf to the FQDN of the server and the names and IP addresses  are correctly set in the name server (printing would not work without it nor would I reach the web interface at all if that wasn't the case.

I even replaced the local resolver configuration of the base Debian container by my own to ensure that host names and addresses in /etc/hosts are the same.

But something is not really believing me:

```
root@62d5056fc295:/usr/lib/cups/cgi-bin# lpstat -le
Brother_MFC-L8650cdw permanent ipp://localhost/printers/Brother_MFC-L8650cdw ipp://mfc-l8650cdw.mydomain.com/ipp
Brother_MFC-L8650CDW_driverless permanent ipp://localhost/printers/Brother_MFC-L8650CDW_driverless ipp://mfc-l8650cdw..mydomain.com/ipp
Brother_MFC-L8650cdw_Fax permanent ipp://localhost/printers/Brother_MFC-L8650cdw_Fax ipp://mfc-l8650cdw.mydomain.com/ipp/
Brother_MFC_L8650cdw permanent ipp://localhost/printers/Brother_MFC_L8650cdw implicitclass://Brother_MFC_L8650cdw/
PDF permanent ipp://localhost/printers/PDF cups-pdf:/
root@62d5056fc295:/usr/lib/cups/cgi-bin#

```

So where is this localhost coming from? 

Logging shows

```
D [02/May/2023:00:28:12 +0000] [CGI] cgiSetVariable: TOTAL=\"5\"
D [02/May/2023:00:28:12 +0000] [CGI] cgiSetVariable: SERVER_NAME=\"192.168.198.18\"
D [02/May/2023:00:28:12 +0000] [CGI] cgiSetVariable: REMOTE_USER=\"admin\"
D [02/May/2023:00:28:12 +0000] [CGI] cgiSetVariable: CUPS_VERSION=\"CUPS v2.3.3op2\"
D [02/May/2023:00:28:12 +0000] [CGI] cgiSetArray: printer_state[0]=\"3\"
D [02/May/2023:00:28:12 +0000] [CGI] cgiSetArray: printer_state_message[0]=\"\"
D [02/May/2023:00:28:12 +0000] [CGI] cgiSetArray: admin_uri[0]=\"ipps://localhost/printers/Brother_MFC-L8650cdw\"
D [02/May/2023:00:28:12 +0000] [CGI] cgiSetArray: printer_uri_supported[0]=\"ipps://localhost/printers/Brother_MFC-L8650cdw\"

```
that someone resolved the "printserv.mydomain.com" to (one of) its (correct) IP addresses bit the admin_uri and printer_uri_supported are not containing this address and the administration pages are using these addresses.

I must have been doing something fundamentally wrong but I'm completely out of clues here... And I can't be the only one ever messing this up.

**Comment by michaelrsweet** [comment]
The URIs reported back to the programs are going to have "localhost" because the programs are communicating locally with the scheduler (cupsd) - over the domain socket. The server name is what gets advertised for *remote* clients...

The CGIs should be rewriting the URIs to provide relative URLs in the web interface (to use the same hostname, port, and scheme) but if that isn't working then you've found a bug.

**Reply by noseshimself** [reply]
I'm still not completely convinced that it is a bug yet...

To be sure that the CGIs will find at least something in their environment I added ``SERVER_NAME ="printserv.mydomain.com"`` to the environment. Yet (see above) the IP address is showing up so the caller explicitly set that. DNS (forward and backward) are ok so if it tried to resolve it that would have worked.

cupsd.conf contains
```
Listen *:443
Listen *:631
Listen /run/cups/cups.sock

ServerName  printserv.sub.mydomain.net
ServerAlias printserv.mydomain.net
ServerAlias printserv.nebula.mydomain.net
```
so the entire subsystem should know its own server name for handing it over to the CGIs.

Is there anything else I should poke my fingers into before opening an issue?
